base_layer/core/src/mining/miner.rs

```diff
@@ -34,18 +35,31 @@ use crate::{
     },
 };
 use core::sync::atomic::AtomicBool;
+use futures::{
+    channel::{
+        mpsc,
+        mpsc::{Receiver, Sender},
+    },
+    future::FutureExt,
+    pin_mut,
+    stream::Fuse,
+    StreamExt,
+};
 use std::{
     sync::{atomic::Ordering, Arc},
     thread,
     time,
 };
+use tari_broadcast_channel::Subscriber;
 use tari_crypto::keys::SecretKey;
+use tokio::runtime;
 
 pub struct Miner<B: BlockchainBackend> {
     kill_flag: Arc<AtomicBool>,
-    stop_mining_flag: Arc<AtomicBool>,
+    got_new_block_event_flag: Arc<AtomicBool>,
```

Comment:
Maybe `received_new_block_event_flag` is a better name.

Reply:
or `received_new_block_flag`

Reply:
done